procrastinate/contrib/django/django_connector.py

```diff
@@ -0,0 +1,207 @@
+    def _execute_query_connection(
+        self, query: str, connection: DatabaseWrapper, **arguments: Any
+    ) -> None:
+        with connection.cursor() as cursor:
+            cursor.execute("LISTEN procrastinate_any_queue;")  # TODO
+            # cursor.execute(
+            #     PERCENT_PATTERN.sub("%%", query), self._wrap_json(arguments)
+            # )
+
+    @wrap_exceptions
+    @sync_to_async
+    def execute_query_one_async(self, query: str, **arguments: Any) -> dict[str, Any]:
+        return self.execute_query_one(query, **arguments)
+
+    @wrap_exceptions
+    def execute_query_one(self, query: str, **arguments: Any) -> dict[str, Any]:
+        with self._connection() as connection:
+            with connection.cursor() as cursor:
+                cursor.execute(
+                    PERCENT_PATTERN.sub("%%", query), self._wrap_json(arguments)
+                )
+                return self._dict_fetchone(cursor)
+
+    @wrap_exceptions
+    @sync_to_async
+    def execute_query_all_async(
+        self, query: str, **arguments: Any
+    ) -> list[dict[str, Any]]:
+        with self._connection() as connection:
+            with connection.cursor() as cursor:
+                cursor.execute(
+                    PERCENT_PATTERN.sub("%%", query), self._wrap_json(arguments)
+                )
+                return self._dict_fetchall(cursor)
+
+    async def listen_notify(
+        self, event: asyncio.Event, channels: Iterable[str]
+    ) -> None:
+        while True:
+            with self._connection() as connection:
+                for channel_name in channels:
+                    await self._execute_query_connection(
+                        query=sql.queries["listen_queue"],
+                        connection=connection,
+                        channel_name=channel_name,
+                    )
+                # Initial set() lets caller know that we're ready to listen
+                event.set()
+                await self._loop_notify(event=event, connection=connection)
+
+    @sync_to_async
+    def _get_pg_conn(self, connection: ConnectionProxy) -> Awaitable[pg_conn]:
+        # connection.ensure_connection()
+        return connection.connection
+
+    async def _wait_for_file_descriptor(self, fd: int, timeout: float = None) -> None:
+        loop = asyncio.get_event_loop()
+        future = asyncio.Future()
+        loop.add_reader(fd, future.set_result, None)
+        future.add_done_callback(lambda f: loop.remove_reader(fd))
+        await asyncio.wait_for(future, timeout=timeout)
+
+    @wrap_exceptions
+    async def _loop_notify(
+        self,
+        event: asyncio.Event,
+        connection: ConnectionProxy,
+        timeout: float = LISTEN_TIMEOUT,
+    ) -> None:
+        while True:
+            pg_conn = await self._get_pg_conn(connection)
+            if pg_conn.closed:
+                return
+
+            try:
+                await self._wait_for_file_descriptor(pg_conn.fileno(), timeout)
```

Comment:
Is there a better solution then this?

Reply:
I think that up until now, we've considered that the listen-notify took a dedicated connection. If we're trying to shoehorn everything into a single connection, I'm afraid there might be consequences. What about, as a first step, disabling listen-notify when using a Django connector ?

Reply:
Actually I need listen-notify in workers in my project. Disabling should be possible with [this flag](https://procrastinate.readthedocs.io/en/stable/reference.html#procrastinate.App.run_worker).

Also I saw similar approach of using listen-notify in Django [here](https://github.com/gavinwahl/django-postgres-queue/blob/24a5d0297a1b7b8d57d2facb249c0600bed7a95b/dpq/queue.py#L65).

Reply:
I think creating separate aiopg connection just for listen-notify is also an option here.